src/core/pointFeature.js

```diff
@@ -25,7 +27,96 @@ geo.pointFeature = function (arg) {
       m_rangeTree = null,
       m_rangeTreeTime = geo.timestamp(),
       s_data = this.data,
-      m_maxRadius = 0;
+      m_maxRadius = 0,
+      m_clustering = arg.clustering,
+      m_clusterTree = null,
+      m_allData = [],
+      m_lastZoom = null,
+      m_ignoreData = false; // flag to ignore data() calls made locally
+
+  ////////////////////////////////////////////////////////////////////////////
+  /**
+   * Get/Set clustering option
+   *
+   * @returns {geo.pointFeature|boolean}
+   */
+  ////////////////////////////////////////////////////////////////////////////
+  this.clustering = function (val) {
+    if (val === undefined) {
+      return m_clustering;
+    }
+    if (m_clustering && !val) {
+      // Throw out the cluster tree and reset the data
+      m_clusterTree = null;
+      m_clustering = false;
+      s_data(m_allData);
+      m_allData = null;
+    } else if (!m_clustering && val) {
+      // Generate the cluster tree
+      m_clustering = true;
+      m_this._clusterData();
+    }
+    return m_this;
+  };
+
+  ////////////////////////////////////////////////////////////////////////////
+  /**
+   * Generate the clustering tree from positions.  This might be async in the
+   * future.
+   */
+  ////////////////////////////////////////////////////////////////////////////
+  this._clusterData = function () {
+    if (!m_clustering) {
+      // no op
+      return;
+    }
+
+    // set clustering options to default if an options argument wasn't supplied
+    var opts = m_clustering === true ? {radius: 0.01} : m_clustering;
+
+    // generate the cluster tree
+    var position = m_this.position();
+    m_clusterTree = new geo.util.ClusterGroup(opts);
+    m_allData.forEach(function (d, i) {
+      var pt = geo.util.normalizeCoordinates(position(d, i));
+      pt.index = i;
+      m_clusterTree.addPoint(pt);
+    });
+    m_lastZoom = null;
+    m_this._handleZoom(m_this.layer().map().zoom());
+  };
+
+  ////////////////////////////////////////////////////////////////////////////
+  /**
+   * Handle zoom events for clustering.  This keeps track of the last
+   * clustering level, and only regenerates the displayed points when the
+   * zoom level changes.
+   */
+  ////////////////////////////////////////////////////////////////////////////
+  this._handleZoom = function (zoom) {
+    var z = Math.floor(zoom); // get the integer level
+    if (!m_clustering || z === m_lastZoom) {
+      // short cut when there is nothing to do
+      return;
+    }
+
+    m_lastZoom = z;
+
+    var data = m_clusterTree.points(z).map(function (d) {
+      return m_allData[d.index];
+    });
+    m_clusterTree.clusters(z).forEach(function (d) {
```

Comment:
I have some idea what's going on here (just looking at the code, didn't run).  It would be nice to have some descriptive text here here for the future.